test/phoenix_live_view_test.exs

```diff
@@ -295,20 +295,29 @@ defmodule Phoenix.LiveViewUnitTest do
 
   describe "assign_new with assigns" do
     test "tracks changes" do
-      assigns = assign_new(@assigns_changes, :key, fn -> raise "won't be invoked" end)
+      assigns = assign_new(@assigns_changes, :key, fn _ -> raise "won't be invoked" end)
       assert assigns.key == "value"
       refute changed?(assigns, :key)
       refute assigns.__changed__[:key]
 
-      assigns = assign_new(@assigns_changes, :another, fn -> "changed" end)
+      assigns = assign_new(@assigns_changes, :another, fn _ -> "changed" end)
       assert assigns.another == "changed"
       assert changed?(assigns, :another)
 
-      assigns = assign_new(@assigns_nil_changes, :another, fn -> "changed" end)
+      assigns = assign_new(@assigns_nil_changes, :another, fn _ -> "changed" end)
       assert assigns.another == "changed"
       assert changed?(assigns, :another)
       assert assigns.__changed__ == nil
     end
+
+    test "has access to new assigns" do
+      assigns = assign_new(@assigns_changes, :another, fn _ -> "changed" end)
+      |> assign_new(:and_another, fn %{another: another} -> another end)
```

Comment:
```suggestion
      assigns =
        @assigns_changes
        |> assign_new(:another, fn _ -> "changed" end)
        |> assign_new(:and_another, fn %{another: another} -> another end)
```

Reply:
Agreed I should made this additive and kept their prior 0 arity.

Reply:
There are other `assign_new` functinos with `is_function(fun, 0)`guards that _may_ need their new pair `assign_new(...) when is_function(fun, 1)`. You can double check by playing with existing tests and passing fun/1 to them.

Reply:
Thought about only providing the fun/1 when using assigns only and not a socket at first?

I ask because I'm unsure about the right approch to. 
```
 %{private: %{assign_new: {assigns, keys}}} ->
        # It is important to store the keys even if they are not in assigns
        # because maybe the controller doesn't have it but the view does.
        socket = put_in(socket.private.assign_new, {assigns, [key | keys]})
        Phoenix.LiveView.Utils.force_assign(socket, key, Map.get_lazy(assigns, key, fun))
```
In the context of using a socket.

Reply:
That's a API design decision so it's not really up to me to say for sure, but as a LV user I'd expect it to just work when passing either a `socket` or `assigns` considering the first arg of `assign_new` is `socket_or_assigns`.

I guess you're asking about the line `Phoenix.LiveView.Utils.force_assign(socket, key, Map.get_lazy(assigns, key, fun))`, you can just write it as:

```elixir
Phoenix.LiveView.Utils.force_assign(socket, key, Map.get_lazy(assigns, key, fn -> fun.(socket.assigns) end))
```

But feel free to wait for the feedback of @josevalim or someone from the core team :)

Reply:
I agree that it would be better to also support sockets, I was just unsure about how. I will try this.

Reply:
Just an update, it probably should be `Phoenix.LiveView.Utils.force_assign(socket, key, Map.get_lazy(assigns, key, fn -> fun.(assigns) end))` instead.

Reply:
Yeah as I suspected the `Map.get_lazy` does not like the new arg. 

```
pry(1)>  Map.get_lazy(assigns, key, fun.(assigns))
** (FunctionClauseError) no function clause matching in Map.get_lazy/3

    The following arguments were given to Map.get_lazy/3:

        # 1
        %{existing: "existing-parent"}

        # 2
        :existing3

        # 3
        "existing-parent"

    Attempted function clauses (showing 1 out of 1):

        def get_lazy(map, key, fun) when is_function(fun, 0)

    (elixir 1.13.2) lib/map.ex:491: Map.get_lazy/3
    (stdlib 3.16.1) erl_eval.erl:685: :erl_eval.do_apply/6
    (iex 1.13.2) lib/iex/evaluator.ex:310: IEx.Evaluator.handle_eval/3
    (iex 1.13.2) lib/iex/evaluator.ex:285: IEx.Evaluator.do_eval/3
    (iex 1.13.2) lib/iex/evaluator.ex:274: IEx.Evaluator.eval/3
    (iex 1.13.2) lib/iex/evaluator.ex:169: IEx.Evaluator.loop/1
    (iex 1.13.2) lib/iex/evaluator.ex:32: IEx.Evaluator.init/4
    (phoenix_live_view 0.17.9) lib/phoenix_live_view.ex:764: Phoenix.LiveView.assign_new/3
    test/phoenix_live_view_test.exs:316: Phoenix.LiveViewUnitTest."test assign_new with socket has access to assigns"/1
    (ex_unit 1.13.2) lib/ex_unit/runner.ex:500: ExUnit.Runner.exec_test/1
    (stdlib 3.16.1) timer.erl:166: :timer.tc/1
    (ex_unit 1.13.2) lib/ex_unit/runner.ex:451: anonymous fn/4 in ExUnit.Runner.spawn_test_monitor/4
 ```

Reply:
Yep `Map.get_lazy/3` expect a fun 👍🏻 

And a test like this should pass:

```elixir
    test "with 1 arity function" do
      socket =
        @socket
        |> assign_new(:another, fn _ -> "changed" end)
        |> assign_new(:and_another, fn %{another: another} -> another end)

      assert socket.assigns == %{
               another: "changed",
               and_another: "changed",
               live_action: nil,
               flash: %{},
               __changed__: %{another: true, and_another: true}
             }
    end
```

Reply:
Before I continue I think its worth discussing the expecting behavior for the socket.  
in the case of https://github.com/phoenixframework/phoenix_live_view/blob/46132e5a8b1ab81f612e5b59eecab53b30da5a06/test/phoenix_live_view_test.exs#L277-L294

You end up here. https://github.com/phoenixframework/phoenix_live_view/blob/46132e5a8b1ab81f612e5b59eecab53b30da5a06/lib/phoenix_live_view.ex#L753-L757

In which case the assigns on the socket have yet to be updated and the new assigns are stored in the private's assign_new.

IE. If I inspect the socket I will see `existing` had not changed while the private's new assigns for existing has. 
```
  assigns: %{
    __changed__: %{existing: true, existing2: true},
    existing: "existing-parent",
    existing2: "existing2",
    flash: %{},
    live_action: nil
  },
  
  vs in sockets private.
  
  %{
  __changed__: %{},
  assign_new: {%{existing: "existing-parent"}, [:existing]},
  connect_info: %{},
  connect_params: %{},
  root_view: Phoenix.LiveViewTest.ParamCounterLive
}
```

Whats the best approach to this?

Reply:
Just to make sure I follow your question.

For the case of the following calls:

```elixir
       |> assign_new(:existing, fn assigns -> "new-existing" end) 
       |> assign_new(:existing2, fn assigns -> "new-existing2" end) 
```

None of those two functions will be called because `:existing` and `:existing2` keys are already declared, so it's fine.

I guess your question is about the last one where the fn would be called passing something as argument:

```elixir
|> assign_new(:notexisting, fn assigns -> "new-notexisting" end)
```

So your concern is what `assigns` should return?



> I guess something like this would be expected?

The expected assigns on that example, IMO should be:

```elixir
existing: "existing-parent", # parent already defined, do not overwrite
existing2: "existing2", # defined on assign/2
existing_parent: "existing-parent" # current value of `existing`
existing3: "existing2" # current value of `existing2`
````

_And no `existing_parent_updated`_

The problem is currently you only see the `existing` assign which is defined on `@socket.private[:assign_new]` so it won't work. _If_ that's the correct expected result then it needs a bit of extra logic to fetch all assigns before passing to `fn -> fun.(assigns) end`. Makes sense?

Sorry I can't say for sure if that's the expected behavior and if it should be implemented, it's just what I'd expect If calling that function. So it's better to wait for some feedback, ok?

Reply:
> IMO the expected assigns on that example should be:
> 
> ```elixir
> existing: "existing-parent", # parent already defined, do not overwrite
> existing2: "existing2", # defined on assign/2
> existing_parent: "existing-parent" # current value of `existing`
> existing3: "existing2" # current value of `existing2`
> ```

Ah I think I miss read the test assuming ` |> assign_new(:existing, fn -> "new-existing" end)` overriding the prior `existing-parent` and I would have access to that.